Documentation/botr/readytorun-format.md

```diff
@@ -4,10 +4,70 @@ ReadyToRun File Format
+
+* The composite R2R file forms a tightly-bound version bubble where an individual assembly cannot be accidentally replaced on disk so we don't need to emit any type layout checks.
+* Similarly, the runtime doesn't need to check reference assembly MVID's within the tightly-bound version bubble represented by the composite file.
+* The composite R2R file can / should share generic instantiations. This both saves space and improves perf as we should be able to emit more instantiations i.o.w. avoid their jitting.
+
+### Structural format changes specific to composite build mode
+
+When a R2R file has the `READYTORUN_FLAG_COMPOSITE` bit set in the R2R header, it conforms to the composite file format. The main differences between composite and single-input R2R PE files are as follows:
+
+* In single-input R2R PE files the COFF header points at the COR header of the input MSIL metadata. In composite R2R files there are multiple input MSIL metadata so that the COFF header instead points at the COR header and metadata representing the manifest metadata and the "actual" input assemblies are tracked under the new R2R header table `READYTORUN_SECTION_ASSEMBLIES`.
+
+* In single-input R2R PE files compiled in the large version bubble mode, the metadata manifest contains the set of assemblies containing functions called from the input assembly (possibly with some transitivity due to inlining); in composite R2R PE files the manifest metadata is supposed to contain all assemblies contained within the composite file. The linear indices of the assemblies in the manifest metadata AssemblyRef table correspond to indexing within the `READYTORUN_SECTION_ASSEMBLIES` and `READYTORUN_SECTION_SIMPLE_ASSEMBLY_NAME_LOOKUP` tables.
+
+* In single-input R2R PE files the R2R section `READYTORUN_SECTION_AVAILABLE_TYPES` encodes a native hashtable of types within a given assembly. In composite R2R files, the available types need to be split per input assembly and are referenced from `READYTORUN_SECTION_ASSEMBLIES`.
+
+* In single-input R2R PE files the R2R section `READYTORUN_SECTION_METHOD_ENTRYPOINTS` encodes a native hashtable of methods within a given assembly. In composite R2R files, the method entrypoints need to be split per input assembly and are referenced from `READYTORUN_SECTION_ASSEMBLIES`.
+
+* In single-input R2R PE files the  generic instantiations generated for the input MSIL assembly are placed in `READYTORUN_SECTION_INSTANCE_METHOD_ENTRYPOINTS`. In composite R2R files, this section represents all instance entrypoints emitted within the composite build (i.e. generic instantiations needed by any of the input assemblies). With the proposed change to make the "input MSIL" represent the manifest metadata, we should be able to easily refer to all required reference assemblies in the signature encodings even without the manifest module override (**note:** this is a minute technical detail).
+
+### Required CoreCLR runtime changes
```

Comment:
Nit: This does not belong to the file format spec.

Reply:
As I mentioned in the introduction, this first version was a bit of a mess due to combining two kinds of docs. After writing up some of that things became somewhat clearer to me. For now I have added a secondary doc readytorun-composite-format-design I seeded with the "design notes" parts I removed from the format md.